src/java.base/aix/native/libnio/MappedMemoryUtils.c

```diff
@@ -0,0 +1,210 @@
+    int result = 0;
+    long i = 0;
+    void *a = (void *) jlong_to_ptr(address);
+    mincore_vec_t* vec = NULL;
+
+    /* See JDK-8186665 */
+    size_t pagesize = (size_t)sysconf(_SC_PAGESIZE);
+    if ((long)pagesize == -1) {
+        return JNI_FALSE;
+    }
+    numPages = (jlong) calculate_number_of_pages_in_range(a, len, pagesize);
+
+    /* Include space for one sentinel byte at the end of the buffer
+     * to catch overflows. */
+    vec = (mincore_vec_t*) malloc(numPages + 1);
+
+    if (vec == NULL) {
+        JNU_ThrowOutOfMemoryError(env, NULL);
+        return JNI_FALSE;
+    }
+
+    vec[numPages] = '\x7f'; /* Write sentinel. */
+    result = mincore(a, (size_t)len, vec);
+    assert(vec[numPages] == '\x7f'); /* Check sentinel. */
+
+    if (result == -1) {
+        JNU_ThrowIOExceptionWithLastError(env, "mincore failed");
+        free(vec);
+        return JNI_FALSE;
+    }
+
+    for (i=0; i<numPages; i++) {
+        if (vec[i] == 0) {
+            loaded = JNI_FALSE;
+            break;
+        }
+    }
+    free(vec);
+    return loaded;
+}
+
+
+JNIEXPORT void JNICALL
+Java_java_nio_MappedMemoryUtils_load0(JNIEnv *env, jobject obj, jlong address,
+                                     jlong len)
+{
+    char *a = (char *)jlong_to_ptr(address);
+    int result = madvise((caddr_t)a, (size_t)len, MADV_WILLNEED);
+    if (result == -1) {
+        JNU_ThrowIOExceptionWithMessageAndLastError(env, "madvise with advise MADV_WILLNEED failed");
+    }
+}
+
+JNIEXPORT void JNICALL
+Java_java_nio_MappedMemoryUtils_unload0(JNIEnv *env, jobject obj, jlong address,
+                                     jlong len)
+{
+    char *a = (char *)jlong_to_ptr(address);
+    int result = madvise((caddr_t)a, (size_t)len, MADV_DONTNEED);
+    if (result == -1) {
+        JNU_ThrowIOExceptionWithMessageAndLastError(env, "madvise with advise MADV_DONTNEED failed");
+    }
+}
+
+static void set_error_if_shared(JNIEnv* env, prmap_t* map_entry)
+{
+    if (map_entry->pr_mflags & MA_SHARED) {
+        // MA_SHARED => MAP_SHARED => !MAP_PRIVATE. This error is valid and should be thrown.
+        JNU_ThrowIOExceptionWithMessageAndLastError(env, "msync with parameter MS_SYNC failed (MAP_SHARED)");
+        return;
+    } else {
+        // O.W. MAP_PRIVATE or no flag was specified and EINVAL is the expected behaviour.
+        return;
+    }
+}
+
+static void check_proc_map_array(JNIEnv* env, FILE* proc_file, prmap_t* map_entry, void* end_address)
+{
+    while (!feof(proc_file)) {
+        memset(map_entry, '\0', sizeof(prmap_t));
+        fread((char*)map_entry, sizeof(prmap_t), 1, proc_file);
+        if (ferror(proc_file)) {
+            JNU_ThrowIOExceptionWithMessageAndLastError(env,
+                        "msync with parameter MS_SYNC failed (could not read /proc/<pid>/map)");
+            return;
+        } else if (map_entry->pr_vaddr <= end_address &&
+                   end_address <= map_entry->pr_vaddr + map_entry->pr_size) {
```

Comment:
```
MappedMemoryUtils.c:140:55: error: arithmetic on a pointer to void is a GNU extension [-Werror,-Wgnu-pointer-arith]
                   end_address <= map_entry->pr_vaddr + map_entry->pr_size) {
                                  ~~~~~~~~~~~~~~~~~~~ ^
```

Reply:
Thanks for finding this. I am still building and testing with xlc 16.1, but it's a good idea to add 17.1 to the standard process as well. I'll fix before merging.

Reply:
Thanks for fixing!

Reply:
```
MappedMemoryUtils.c:140:32: error: ordered comparison between pointer and integer ('void *' and 'unsigned long') [-Werror]
                   end_address <= (uint64_t)map_entry->pr_vaddr + map_entry->pr_size) {
                   ~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```